bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java

```diff
@@ -29,211 +34,315 @@
+            throws ConfigurationException, MqttException {
+        String name = brokerConnectionConfig.get("name").toString();
+        if (name == null) {
+            throw new ConfigurationException("MQTT Broker property '{}' should have the format 'broker.propertykey'");
+        }
+        MqttBrokerConnection connection;
+        synchronized (brokerConnections) {
+            connection = brokerConnections.get(name.toLowerCase());
+            if (connection == null) {
+                connection = new MqttBrokerConnection(name, brokerConnectionConfig.get("url").toString());
+                brokerConnections.put(name.toLowerCase(), connection);
+                for (MqttBrokersObserver o : brokersObservers) {
+                    o.brokerAdded(connection);
+                }
+            }
+        }
+
+        connection.setCredentials(brokerConnectionConfig.get("user").toString(),
+                brokerConnectionConfig.get("pwd").toString());
+        connection.setQos((Integer) brokerConnectionConfig.get("qos"));
+        connection.setRetain((Boolean) brokerConnectionConfig.get("retain"));
+        connection.setClientId(brokerConnectionConfig.get("clientId").toString());
+        connection.setKeepAliveInterval((Integer) brokerConnectionConfig.get("keepAlive"));
+        MqttWillAndTestament will = MqttWillAndTestament.fromString(brokerConnectionConfig.get("lwt").toString());
+        if (will != null) {
+            logger.debug("Setting last will: {}", will);
+            connection.setLastWill(will);
         }
+        return connection;
     }
 
     /**
-     * Lookup an broker connection by name.
+     * Remove a broker connection
      *
-     * @param brokerName to look for.
-     * @return existing connection or new one if it didn't exist yet.
+     * @param connection The broker connection
      */
-    private MqttBrokerConnection getConnection(String brokerName) {
+    public void removeBrokerConnection(MqttBrokerConnection connection) {
         synchronized (brokerConnections) {
-            MqttBrokerConnection conn = brokerConnections.get(brokerName.toLowerCase());
-            if (conn == null) {
-                conn = new MqttBrokerConnection(brokerName);
-                brokerConnections.put(brokerName.toLowerCase(), conn);
+            brokerConnections.remove(connection.getName(), connection);
+            for (MqttBrokersObserver o : brokersObservers) {
+                o.brokerRemoved(connection);
             }
-            return conn;
         }
     }
 
     /**
+     * Remove a broker connection by name
+     *
+     * @param brokerName The broker name
+     */
+    public void removeBrokerConnection(String brokerName) {
+        synchronized (brokerConnections) {
+            MqttBrokerConnection connection = brokerConnections.get(brokerName.toLowerCase());
+            if (connection == null) {
+                return;
+            }
+            brokerConnections.remove(brokerName);
```

Comment:
here again `brokerName` may not be equal to `brokerName.toLowerCase()` in which case the observers get notified but the connection is still in the map. You may call `brokerConnections.remove(brokerName.toLowerCase(), connection)` here and again check the return value. If the remove fails you may log.debug to give a hint what happened.